File: src/project/projectidentity.cpp
#include "projectidentity.h"

#include <QDir>
#include <QHash>

#include <algorithm>

std::optional<SongName> SongName::create(QString value)
{
    if (value.isEmpty())
        return std::nullopt;
    return SongName{std::move(value)};
}

SongName::SongName(QString value) : m_value{std::move(value)} {}

size_t qHash(const SongName &name, size_t seed)
{
    return qHash(name.value(), seed);
}

std::optional<VoicegroupId> VoicegroupId::create(QString sourceRelativePath, QString sectionLabel)
{
    // Normalize the spelling ("./a", "a//b") while rejecting anything that
    // does not name a file inside the project: an absolute path, a
    // ".."-escape, or the project root itself (any spelling that cleans
    // down to ".").
    const QString normalized = QDir::cleanPath(sourceRelativePath);
    if (normalized.isEmpty() || normalized == QLatin1String(".") ||
        QDir::isAbsolutePath(normalized) || normalized == QLatin1String("..") ||
        normalized.startsWith(QLatin1String("../")))
        return std::nullopt;
    return VoicegroupId{normalized, std::move(sectionLabel)};
}

VoicegroupId::VoicegroupId(QString normalizedSourceRelativePath, QString sectionLabel)
    : m_sourceRelativePath{std::move(normalizedSourceRelativePath)}
    , m_sectionLabel{std::move(sectionLabel)}
{}

size_t qHash(const VoicegroupId &id, size_t seed)
{
    return qHashMulti(seed, id.sourceRelativePath(), id.sectionLabel());
}

SavedWorkspaceRecipe normalizeSavedRecipe(QString projectPath, QStringList labels,
                                          QString selectedLabel)
{
    QVector<SongName> ordered;
    for (QString &label : labels) {
        auto name = SongName::create(std::move(label));
        if (!name || std::find(ordered.cbegin(), ordered.cend(), *name) != ordered.cend())
            continue;
        ordered.push_back(std::move(*name));
    }

    auto selected = SongName::create(std::move(selectedLabel));
    if (selected && ordered.isEmpty()) {
        // Pre-tabs session generation: no tab list survived discarding, so
        // the selected song restores as the one tab.
        ordered.push_back(*selected);
    } else if (!selected ||
               std::find(ordered.cbegin(), ordered.cend(), *selected) == ordered.cend()) {
        // The selected label is empty, missing from the ordered list, or
        // discarded: fall back to the first name.
        selected = ordered.isEmpty() ? std::nullopt : std::optional<SongName>{ordered.front()};
    }
    return SavedWorkspaceRecipe{std::move(projectPath), std::move(ordered), std::move(selected)};
}

File: src/swift/project/ProjectIdentity.swift
/// A nonempty song label, used as a stable project-relative identity.
public struct SongName: Hashable, Sendable {
    /// The song label exactly as supplied.
    public let value: String

    /// Creates an identity unless the label is empty.
    /// - Parameter value: The song label.
    public init?(_ value: String) {
        guard !value.isEmpty else { return nil }
        self.value = value
    }
}

/// A voicegroup source path and its optional, unmodified section label.
public struct VoicegroupId: Hashable, Sendable {
    /// The normalized project-relative source path.
    public let sourceRelativePath: String
    /// The section label exactly as supplied.
    public let sectionLabel: String

    /// Creates an identity only for a source file inside the project.
    /// - Parameters:
    ///   - sourceRelativePath: The project-relative source path to normalize.
    ///   - sectionLabel: The section label, which is not normalized.
    public init?(sourceRelativePath: String, sectionLabel: String) {
        let normalized = ProjectFileStore.cleanPath(sourceRelativePath)
        guard !normalized.isEmpty,
              normalized != ".",
              !ProjectFileStore.isAbsolutePath(normalized),
              normalized != "..",
              !normalized.hasPrefix("../")
        else { return nil }
        self.sourceRelativePath = normalized
        self.sectionLabel = sectionLabel
    }
}

/// A saved workspace after duplicate and invalid labels have been discarded.
public struct SavedWorkspaceRecipe: Sendable {
    /// The project path exactly as supplied.
    public let projectPath: String
    /// Distinct, nonempty song labels in their original order.
    public let orderedSongs: [SongName]
    /// The selected song, if any valid song survived.
    public let selected: SongName?
}

/// Restores ordered song tabs and a valid selected song from saved labels.
/// - Parameters:
///   - projectPath: The project path to preserve unchanged.
///   - labels: The saved ordered song labels.
///   - selected: The saved selected song label.
/// - Returns: The normalized workspace recipe.
public func normalizeSavedRecipe(projectPath: String, labels: [String], selected: String) -> SavedWorkspaceRecipe {
    var orderedSongs: [SongName] = []
    var seen: Set<SongName> = []
    for label in labels {
        guard let name = SongName(label), seen.insert(name).inserted else { continue }
        orderedSongs.append(name)
    }

    var selectedSong = SongName(selected)
    if let candidate = selectedSong {
        if orderedSongs.isEmpty {
            orderedSongs.append(candidate)
        } else if !seen.contains(candidate) {
            selectedSong = orderedSongs.first
        }
    } else {
        selectedSong = orderedSongs.first
    }
    return SavedWorkspaceRecipe(projectPath: projectPath, orderedSongs: orderedSongs, selected: selectedSong)
}
